WIP Add Robust Multi-GPU and Cross-Device Support for PyIsolate Benchmarks

billybasass opened the pull request:
This PR adds, cross-vendor GPU support to the PyIsolate benchmarking suite, enabling seamless benchmarking on NVIDIA (CUDA), AMD (ROCm), and Intel (XPU/oneAPI) GPUs, as well as CPU. It also improves device selection, serialization, and reporting for all supported platforms.
Key Features & Changes
Multi-GPU Device Selection:
Detects all available GPUs (NVIDIA, AMD, Intel) in the system.
Prompts the user to select which GPU to use if more than one is present.
Prints the name of the selected device at the start of each benchmark run.
Device-Aware Tensor Serialization:
CPU and CUDA: Uses PyTorch’s native (zero-copy) serialization for inter-process communication.
Intel XPU: Implements a fallback serialization for XPU tensors:
XPU tensors are moved to CPU, serialized as a buffer (with shape and dtype), sent through the multiprocessing queue, and reconstructed as XPU tensors on the receiving side.
This avoids PyCapsule/pickle errors and ensures all tests run on Intel GPUs, even though the transfer is not zero-copy.
All benchmarks (including shared-torch and RPC) now run on Intel GPUs.
Benchmark Script Enhancements:
Benchmarks are run on all available backends (CPU, CUDA, XPU).
Status messages and results include the backend and device name for clarity and reproducibility.
Extension Venv and Dependency Handling:
Extension venvs now install the correct PyTorch wheel and index for each backend.
Ensures correct torch version and index for Intel GPUs, and removes invalid version specifiers.
Linter and Syntax Fixes:
Fixed all linter and syntax errors, including stray else: statements and indentation issues.
Codebase is now linter-clean and CI-ready.
Motivation
Enable robust benchmarking and extension support across all major GPU vendors.
Ensure that all tests run on all supported hardware, with clear device selection and reporting.
Provide a future-proof serialization mechanism for device tensors, especially for Intel XPU, until native support is available in PyTorch.
Testing
Benchmarks were run on systems with NVIDIA, AMD, and Intel GPUs.
All tests pass, and results are correctly reported for each device.
No PyCapsule/pickle errors on Intel GPUs.
Closes Issues
Fixes device selection and reporting for multi-GPU systems.
Fixes XPU tensor serialization for multiprocessing.
Fixes linter and syntax errors in the codebase.
Ready for review!

billybasass committed:
Add multi-GPU, Intel/AMD support, robust device selection, and linter fixes

billybasass committed:
main change is device type was a integer and changed it to a string some you can still pick cpu only testing

billybasass committed:
fix redundant torch requirements and correct wheels

billybasass committed:
set python to 3.12 because 3.13 does not have all wheels and make the array writable before converting to a tensor

billybasass committed:
ruff and lint checks

billybasass committed:
fix: venv isolation for extension imports, ensure only venv site-packages are used; codebase ruff clean

billybasass committed:
intel testing added

billybasass committed:
venv isolation has been applied

billybasass committed:
added GPU share results for intel GPUs

billybasass committed:
The memory benchmark will now loop over all available backends (including xpu for Intel GPUs)

billybasass committed:
The memory benchmark will now always run with --backend auto and test loop to always include CPU in the results, regardless of GPU presence.

billybasass committed:
Refactor: make GPU features optional, move GPU utils to gpu_utils.py, added back pre commit config, and clean venv handling

billybasass committed:
Ruff test removed name so I added it back and All device selection and environment setup is now handled in main(). run_benchmarks() is now device-agnostic and will not throw UnboundLocalError

guill commented on pyisolate/_internal/host.py:
```diff
@@ -340,32 +342,59 @@ def _install_dependencies(self):
                 safe_dependencies.append(dep)
 
+            # Only add torch requirement if not already present
+            if torch_requirement and not torch_in_deps:
+                safe_dependencies.insert(0, torch_requirement)
+                uv_args += torch_index_args
```
If we have duplicate incompatible torch requirements (because the user specified a version of torch different than the one provided by `share_torch`), that *should* result in a failure. We need to guarantee that we're using the exact same version of torch in the host and child process or we may experience memory corruption.

guill commented on pyisolate/_internal/shared.py:
```diff
@@ -407,47 +422,49 @@ def _send_thread(self):
-                            pending["calling_loop"].call_soon_threadsafe(
-                                pending["future"].set_exception,
-                                RuntimeError(f"CUDA out of memory during request transmission: {error_msg}"),
+                        try:
+                            simple_response = RPCRequest(
+                                kind="call",
+                                object_id=item["object_id"],
+                                call_id=call_id,
+                                parent_call_id=item["parent_call_id"],
+                                method=item["method"],
+                                args=(),
+                                kwargs={},
                             )
+                            self.send_queue.put(simple_response)
+                        except Exception:
+                            print("Failed to send even a simple error request - process may be stuck")
```
I'm not really sure what's going on here. If there's an exception in sending a request, we can't just try to send it again -- that could result in duplicate calls. It also means that the calling process is blocked permanently (since it's waiting on a response that will never come). We need the host process to be resilient to failures in the child processes which we lose with this change.

guill commented on tests/test_benchmarks.py:
```diff
@@ -27,6 +28,18 @@
+            f"{torch.cuda.get_device_name(torch.cuda.current_device())}"
+        )
+    else:
+        print("[PyIsolate] CUDA not available, using CPU only.")
```
Similar to previous comments, let's just fail in this case instead of silently falling back to CPU-only and succeeding.

guill commented on tests/test_benchmarks.py:
```diff
@@ -191,135 +204,36 @@ def print_summary(self):
 class TestRPCBenchmarks(IntegrationTestBase):
     """Benchmark tests for RPC call overhead."""
 
+    benchmark_ext_shared: Optional[object] = None
```
Can we be more specific about the type here than just `object`?

guill commented on tests/test_benchmarks.py:
```diff
@@ -336,6 +250,7 @@ async def test_small_data_benchmarks(self):
         print("SMALL DATA BENCHMARKS")
         print("=" * 60)
 
+        assert self.runner is not None  # type: ignore
```
Why do we need `# type: ignore` here?

guill reviewed (changes requested):
A number of changes requested. Put into broad categories, they are generally related to:

1. Type hint warnings that have been suppressed but should probably just be fixed to not need suppression.
2. Silently falling back to CPU-only rather than failing a test if we e.g. are trying to test CUDA but don't have CUDA installed.
3. Various one-off changes.

billybasass committed:
[200~Removed unnecessary getattr(..., 'manager', None) checks; now directly uses test_instance.manager.

billybasass committed:
Updated device selection logic to allow tests/benchmarks to run on Intel XPU and AMD ROCm GPUs, not just CUDA.Fixed unresolved import for ExampleExtensionBase by importing from example.shared.

billybasass commented:
Thank you for the great reply @guill, some of the unneeded things there left over for when I tried a different approach you suggested and I thought I got rid of (working under the weather) thank you for finding them. There is more than one place you did not know why I was using type ignore, it was because somethings were annotated and some not annotated so Python treats it as dynamic, and no error is raise, I have fixed this. You asked what ruff I used, that I updated to the lastest before running. For the hardcoding of pytorch when it comes to Intel GPUs. This should only happen with Intel GPUs to use pytorch 2.7 or up and is needed because Intel was not fully integrated with pytorch tell 2.7. I did ask Bob Duffy from Intel about this and we will probably have problems if we let it run lower. 

Here is a summary of changes I made after checking every line you pointed out.

1. Type hint warnings and suppressions
Updated type annotations for benchmark_ext and benchmark_ext_shared to use Optional example.shared instead of object or untyped.
Removed unnecessary # type: ignore suppressions for these assignments and for runner assertions in tests.
Updated imports to include example.shared where needed.

2. Failing on missing GPU/ROCm instead of falling back to CPU
In both benchmarks/benchmark.py and benchmarks/memory_benchmark.py, the code now calls sys.exit(1) if no supported GPU backend is available, or if ROCm is not supported on Windows, instead of silently falling back to CPU.
In tests/test_benchmarks.py, the test now fails (exits) if CUDA is not available, rather than printing a warning and continuing.

3. One-off changes and cleanups
Removed unnecessary getattr(..., 'manager', None) checks; now directly uses test_instance.manager.
Removed unnecessary type: ignore[attr-defined] suppressions for NVML calls in memory_benchmark.py.
Removed redundant explicit float casts where Python's type inference is sufficient.
Removed unnecessary getattr(self.test_base, "test_root", None) checks, using self.test_base.test_root directly.
Cleaned up comments and added clarifying comments where appropriate.
Removed tabulate from the [project.optional-dependencies.test] section in pyproject.toml, keeping it only in [bench].

pollockjj commented:
No longer able to merge due to code divergence.
